pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -137,3 +135,81 @@ func (r *DNSPolicyReconciler) deleteGatewayDNSRecords(ctx context.Context, gatew
 	}
 	return nil
 }
+
+func (r *DNSPolicyReconciler) buildClusterGateway(ctx context.Context, downstreamClusterName string, clusterAddress []gatewayv1beta1.GatewayAddress) (dns.ClusterGateway, error) {
+	var target dns.ClusterGateway
+	singleClusterAddresses := make([]gatewayv1beta1.GatewayAddress, len(clusterAddress))
+
+	mc := &clusterv1.ManagedCluster{}
+	if err := r.Client().Get(ctx, client.ObjectKey{Name: downstreamClusterName}, mc, &client.GetOptions{}); err != nil {
+		return target, err
+	}
+
+	for i, addr := range clusterAddress {
+		addrType := gatewayv1beta1.IPAddressType
+		if *addr.Type == gateway.MultiClusterHostnameAddressType {
+			addrType = gatewayv1beta1.HostnameAddressType
+		}
+
+		singleClusterAddresses[i] = gatewayv1beta1.GatewayAddress{
+			Type:  &addrType,
+			Value: addr.Value,
+		}
+	}
+
+	target = *dns.NewClusterGateway(mc, singleClusterAddresses)
+
+	return target, nil
+}
+
+func getClusterGatewayAddresses(gw *gatewayv1beta1.Gateway) map[string][]gatewayv1beta1.GatewayAddress {
+	clusterAddrs := make(map[string][]gatewayv1beta1.GatewayAddress, len(gw.Status.Addresses))
+
+	for _, address := range gw.Status.Addresses {
+		var gatewayAddresses []gatewayv1beta1.GatewayAddress
+
+		//Default to Single Cluster (Normal Gateway Status)
+		cluster := "none"
+		addressValue := address.Value
+
+		//Check for Multi Cluster (MGC Gateway Status)
+		if *address.Type == gateway.MultiClusterIPAddressType || *address.Type == gateway.MultiClusterHostnameAddressType {
+			tmpCluster, tmpAddress, found := strings.Cut(address.Value, "/")
+			//If this fails something is wrong and the value hasn't been set correctly
+			if found {
+				cluster = tmpCluster
+				addressValue = tmpAddress
+			}
+		}
+
+		gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
+			Type:  address.Type,
+			Value: addressValue,
+		})
+		clusterAddrs[cluster] = gatewayAddresses
+	}
+
+	return clusterAddrs
+}
+
+func listenerTotalAttachedRoutes(upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster string, specListener gatewayv1beta1.Listener, addresses []gatewayv1beta1.GatewayAddress) int {
+	for _, statusListener := range upstreamGateway.Status.Listeners {
+		// assuming all adresses of the same type on the gateway
+		// for Multi Cluster (MGC Gateway Status)
+		if *addresses[0].Type == gateway.MultiClusterIPAddressType || *addresses[0].Type == gateway.MultiClusterHostnameAddressType {
```

Comment:
Not ideal having to pass in all the addresses here just to work out if it's an MGC updated status we are dealing with or not. I can see why you have done it this way though, so maybe a better way of determining the type of Gateway we are dealing with, and the expected status, can be looked into later.

Reply:
adding this to the [stockpile](https://github.com/Kuadrant/multicluster-gateway-controller/issues/496#issuecomment-1736930282) of "fix later"